app/forms/openid_connect_token_form.rb

```diff
@@ -0,0 +1,95 @@
+class OpenidConnectTokenForm
+  include ActiveModel::Model
+  include ActionView::Helpers::TranslationHelper
+  include Rails.application.routes.url_helpers
+
+  attr_reader :grant_type,
+              :code,
+              :client_assertion_type,
+              :client_assertion
+
+  CLIENT_ASSERTION_TYPE = 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer'.freeze
+
+  validates_inclusion_of :grant_type, in: %w(authorization_code)
+  validates_inclusion_of :client_assertion_type,
+                         in: [CLIENT_ASSERTION_TYPE]
+
+  validate :validate_code
+  validate :validate_client_assertion
+
+  def initialize(params)
+    %i(grant_type code client_assertion_type client_assertion).each do |key|
+      instance_variable_set(:"@#{key}", params[key])
+    end
+  end
+
+  def submit
+    {
+      success: valid?,
+      response: response,
+      client_id: client_id,
+      errors: errors.messages
+    }
+  end
+
+  private
+
+  def validate_code
+    errors.add :code, t('openid_connect.token.errors.invalid_code') unless identity.present?
+  end
+
+  def validate_client_assertion
+    return unless identity.present?
+
+    service_provider = ServiceProvider.new(client_id)
+
+    JWT.decode(client_assertion, service_provider.ssl_cert.public_key, true,
+               algorithm: 'RS256',
+               iss: client_id, verify_iss: true,
+               sub: client_id, verify_sub: true,
+               aud: openid_connect_token_url, verify_aud: true)
+  rescue JWT::DecodeError => err
+    # TODO: i18n these JWT gem error messages
+    errors.add(:client_assertion, err.message)
+  end
+
+  def identity
+    # Check if @_identity has already been set (even to nil) to avoid additional DB queries
+    return @_identity if defined?(@_identity)
+    @_identity = Identity.where(session_uuid: code).first
```

Comment:
Got it! I misread the `validate_code` validation and I hadn't noticed the spec with ` before { user.identities.delete_all }`. In that case, the memoization + Reek offense can both be solved by assigning the ivar in the constructor, and making `identity` a private `attr_reader`:
```ruby
def initialize(params)
  %i(grant_type code client_assertion_type client_assertion).each do |key|
    instance_variable_set(:"@#{key}", params[key])
  end

  @identity = Identity.where(session_uuid: code).first
end

private

attr_reader :identity
```
If this performance is important to us, we might want to add a test for this, for example:
```ruby
matching_identities = instance_double(ActiveRecord::Relation)
allow(Identity).to receive(:where).with(session_uuid: code).and_return(matching_identities)
expect(matching_identities).to receive(:first).exactly(:once).and_return(Identity.last)
```